components/emoji_picker/modules/emoji_selector.vue

```diff
@@ -444,13 +478,202 @@ function setTabLabelObserver () {
+  event.preventDefault();
+  hoverFirstEmoji.value = false;
+
+  if (event.key === 'ArrowUp') {
+    const position = indexEmoji % EMOJIS_PER_ROW;
+
+    if (!focusEmoji(0, indexEmoji - EMOJIS_PER_ROW)) {
+      const lastEmojiPosition = emojiFilteredRefs.value.length - (emojiFilteredRefs.value.length % EMOJIS_PER_ROW) + position;
+
+      focusEmoji(0, lastEmojiPosition);
+
+      if (!focusEmoji(0, lastEmojiPosition)) {
+        focusEmoji(0, emojiFilteredRefs.value.length - 1);
+      }
+    }
+  }
+
+  if (event.key === 'ArrowDown') {
+    if (!focusEmoji(0, indexEmoji + EMOJIS_PER_ROW)) {
+      const position = indexEmoji % EMOJIS_PER_ROW;
+
+      if (emojiFilteredRefs.value?.[indexEmoji + (EMOJIS_PER_ROW - position)]) {
+        focusEmoji(0, emojiFilteredRefs.value.length - 1);
+      } else {
+        focusEmoji(0, position);
+      }
+    }
+  }
+
+  if (event.key === 'ArrowLeft') {
+    if (!focusEmoji(0, indexEmoji - 1)) {
+      // Jump to the last emoji of the previous tab
+      // handle end of tab
+      focusEmoji(0, emojiFilteredRefs.value.length - 1);
+    }
+  }
+
+  if (event.key === 'ArrowRight') {
+    if (!focusEmoji(0, indexEmoji + 1)) {
+      focusEmoji(0, 0);
+    }
+  }
+
+  if (event.key === 'Tab') {
+    emits('focus-skin-selector');
+  }
+
+  if (event.key === 'Enter') {
+    emits('selected-emoji', emoji);
+  }
+};
+
+const handleKeyDown = (event, indexTab, indexEmoji, emoji) => {
+  event.preventDefault();
+
+  if (event.key === 'ArrowUp') {
+    const position = indexEmoji % EMOJIS_PER_ROW;
+
+    if (indexTab === 0) {
+      // we are on the first emoji tab, then we should jump to the last row of the last emoji tab
+      const numberOfMissingEmojis = EMOJIS_PER_ROW - (emojiRefs.value[emojiRefs.value.length - 1].length % EMOJIS_PER_ROW);
+
+      const emojiToJump = emojiRefs.value[emojiRefs.value.length - 1].length + numberOfMissingEmojis - (EMOJIS_PER_ROW - position);
+
+      if (!focusEmoji(emojiRefs.value.length - 1, emojiToJump)) {
+        // if there is no emoji in this position, jump to the last emoji of the row
+        focusEmoji(emojiRefs.value.length - 1, emojiRefs.value[emojiRefs.value.length - 1].length - 1);
+      }
+      return;
+    }
+
+    // if we are not on the first tab, we should jump to the previous row of the current tab
+    if (!focusEmoji(indexTab, indexEmoji - EMOJIS_PER_ROW)) {
+      // if there is no previous row, we should jump to emoji in the sampe position of the previous tab
+      const previousTab = indexTab - 1 < 0 ? 0 : indexTab - 1;
+      const emojisInPreviousTab = emojiRefs.value[previousTab].length;
+      const lastEmojiPosition = emojisInPreviousTab - (emojisInPreviousTab % EMOJIS_PER_ROW) + position;
+
+      if (!focusEmoji(previousTab, lastEmojiPosition)) {
+        // if there is no emoji in this position, jump to the last emoji of the row
+        focusEmoji(indexTab - 1, emojiRefs.value[indexTab - 1].length - 1);
+      }
+    }
```

Comment:
We are doing a lot of work in here related to calculating emoji index. Would it make things cleaner if we added a mapping of sorts for the tab start and end indexes?

something like this off the top of my head:

```js
const indexTabMap = [
  // indexTab 0
  { indexStart: 0, indexEnd: 43 },
  // indexTab 1
  { indexStart: 44, indexEnd: 106 },
  ...
]
```

Reply:
I create a composable to handle this in another file, it is not what you expect but its a point of start...

I would like to tackel these on the improvement task..

But as you suggest, I can "hardcode" more our emojiPicker functionality, because we know which emojis have and how many... so.. If we want ... we can do this things more "easy"..

You may ask why I say "hardcode MORE our emojiPicker" the only thing I guess it is necessary to have is the `width: 372px;` on our `.d-emoji-picker{` class, this because we force to have have 9 emoji per row, so all the following calculations will be OK, because we have our constant `EMOJIS_PER_ROW = 9`

Reply:
yeah for sure I think it's fine to have this hardcoded for now. Could be a future upgrade to make width changable if it is actually needed by product for some reason.

And yeah this is not crucial at all for right now, could definitely be a future task